File: model/image_upscale.go
package model

type ImageUpscaleJobStatus string

const (
	ImageUpscaleJobStatusQueued      ImageUpscaleJobStatus = "queued"
	ImageUpscaleJobStatusProcessing  ImageUpscaleJobStatus = "processing"
	ImageUpscaleJobStatusDownloading ImageUpscaleJobStatus = "downloading"
	ImageUpscaleJobStatusSucceeded   ImageUpscaleJobStatus = "succeeded"
	ImageUpscaleJobStatusFailed      ImageUpscaleJobStatus = "failed"
)

type ImageUpscaleJob struct {
	ID                string                `json:"id" gorm:"primaryKey"`
	UserID            string                `json:"-" gorm:"index"`
	ProjectID         string                `json:"projectId" gorm:"index"`
	CanvasID          string                `json:"canvasId" gorm:"index"`
	SourceNodeID      string                `json:"sourceNodeId"`
	SourceAssetID     string                `json:"sourceAssetId"`
	Provider          string                `json:"provider"`
	ProviderRequestID string                `json:"providerRequestId"`
	Model             string                `json:"model"`
	Strategy          string                `json:"strategy"`
	Scale             int                   `json:"scale"`
	Status            ImageUpscaleJobStatus `json:"status" gorm:"index"`
	Progress          int                   `json:"progress"`
	Attempt           int                   `json:"attempt"`
	InputWidth        int                   `json:"inputWidth"`
	InputHeight       int                   `json:"inputHeight"`
	InputMIMEType     string                `json:"inputMimeType"`
	InputBytes        int64                 `json:"inputBytes"`
	InputPath         string                `json:"-" gorm:"type:text"`
	ResultURL         string                `json:"resultUrl" gorm:"type:text"`
	ResultMIMEType    string                `json:"resultMimeType"`
	ResultBytes       int64                 `json:"resultBytes"`
	OutputWidth       int                   `json:"outputWidth"`
	OutputHeight      int                   `json:"outputHeight"`
	ErrorCode         string                `json:"errorCode"`
	ErrorMessage      string                `json:"errorMessage" gorm:"type:text"`
	CloudProcessing   bool                  `json:"cloudProcessing"`
	CreatedAt         string                `json:"createdAt" gorm:"index"`
	StartedAt         string                `json:"startedAt"`
	CompletedAt       string                `json:"completedAt"`
	UpdatedAt         string                `json:"updatedAt"`
}

File: web/src/services/api/image-upscale.ts
import { useUserStore } from "@/stores/use-user-store";
import { apiGet, apiPostEmpty, apiPostForm } from "./request";

export type ImageUpscaleJobStatus = "queued" | "processing" | "downloading" | "succeeded" | "failed";

export type ImageUpscaleCapabilities = {
    enabled: boolean;
    provider: string;
    scales: Array<2 | 4>;
    maxInputBytes: number;
    maxLongEdge: number;
    maxShortEdge: number;
    cloudProcessing: true;
};

export type ImageUpscaleJob = {
    id: string;
    projectId: string;
    canvasId: string;
    sourceNodeId: string;
    sourceAssetId: string;
    provider: string;
    providerRequestId: string;
    model: string;
    strategy: string;
    scale: 2 | 4;
    status: ImageUpscaleJobStatus;
    progress: number;
    attempt: number;
    inputWidth: number;
    inputHeight: number;
    inputMimeType: string;
    inputBytes: number;
    resultUrl: string;
    resultMimeType: string;
    resultBytes: number;
    outputWidth: number;
    outputHeight: number;
    errorCode: string;
    errorMessage: string;
    cloudProcessing: true;
    createdAt: string;
    startedAt: string;
    completedAt: string;
    updatedAt: string;
};

export type CreateImageUpscaleJobInput = {
    file: Blob;
    filename: string;
    scale: 2 | 4;
    projectId: string;
    canvasId: string;
    sourceNodeId: string;
    sourceAssetId?: string;
};

const token = () => useUserStore.getState().token;

export function getImageUpscaleCapabilities() {
    return apiGet<ImageUpscaleCapabilities>("/api/v1/image-upscale/capabilities", undefined, token());
}

export function createImageUpscaleJob(input: CreateImageUpscaleJobInput) {
    const form = new FormData();
    form.append("file", input.file, input.filename);
    form.append("scale", String(input.scale));
    form.append("projectId", input.projectId);
    form.append("canvasId", input.canvasId);
    form.append("sourceNodeId", input.sourceNodeId);
    form.append("sourceAssetId", input.sourceAssetId || "");
    return apiPostForm<ImageUpscaleJob>("/api/v1/image-upscale/jobs", form, token());
}

export function getImageUpscaleJob(jobId: string) {
    return apiGet<ImageUpscaleJob>(`/api/v1/image-upscale/jobs/${encodeURIComponent(jobId)}`, undefined, token());
}

export function retryImageUpscaleJob(jobId: string) {
    return apiPostEmpty<ImageUpscaleJob>(`/api/v1/image-upscale/jobs/${encodeURIComponent(jobId)}/retry`, token());
}
